src/timer/SDL_timer.c

```diff
@@ -201,7 +243,78 @@ static int SDLCALL SDL_TimerThread(void *_data)
            extra times.
          */
-        SDL_WaitSemaphoreTimeoutNS(data->sem, delay);
+        switch (current->callback_data.type) {
+        case SDL_TIMER_CALLBACK_TYPE_MS:
+        case SDL_TIMER_CALLBACK_TYPE_NS:
+            SDL_WaitSemaphoreTimeoutNS(data->sem, delay);
+            break;
+        case SDL_TIMER_CALLBACK_TYPE_PRECISE:
+        {
+            /* This is an adaptation of the algorithm in SDL_DelayPrecise() to
+               produce proper timer behavior, ending the wait if signaled by
+               the semaphore. Notes about how the algorithm works are in
+               SDL_DelayPrecise()'s code.
+             */
+
+            bool signaled = false;
+            const Uint64 SHORT_SLEEP_NS = 1 * SDL_NS_PER_MS;
+            const Uint64 target_value = tick + delay;
+            Uint64 max_sleep_ns = SHORT_SLEEP_NS;
+            while (now + max_sleep_ns < target_value) {
+                if (SDL_WaitSemaphoreTimeoutNS(data->sem, SHORT_SLEEP_NS)) {
+                    signaled = true;
+                    goto precise_timer_done;
+                }
+
+                const Uint64 next_now = SDL_GetTicksNS();
+                const Uint64 next_sleep_ns = (next_now - now);
+                if (next_sleep_ns > max_sleep_ns) {
+                    max_sleep_ns = next_sleep_ns;
+                }
+                now = next_now;
+            }
+
+            if (now < target_value) {
+                if ((target_value - now) > (max_sleep_ns - SHORT_SLEEP_NS)) {
+                    const Uint64 delay_ns = (target_value - now) - (max_sleep_ns - SHORT_SLEEP_NS);
+                    if (SDL_WaitSemaphoreTimeoutNS(data->sem, delay_ns)) {
+                        signaled = true;
+                        goto precise_timer_done;
+                    } else {
+                        now = SDL_GetTicksNS();
+                    }
+                }
+            } else {
+                signaled = SDL_WaitSemaphoreTimeoutNS(data->sem, 0);
+                goto precise_timer_done;
+            }
+
+            while (now + SHORT_SLEEP_NS < target_value) {
+                if (SDL_WaitSemaphoreTimeoutNS(data->sem, SHORT_SLEEP_NS)) {
+                    signaled = true;
+                    goto precise_timer_done;
+                } else {
+                    now = SDL_GetTicksNS();
+                }
+            }
+
+            while (now < target_value) {
+                if (SDL_WaitSemaphoreTimeoutNS(data->sem, 0)) {
+                    signaled = true;
+                    goto precise_timer_done;
+                }
+                now = SDL_GetTicksNS();
+            }
+
+        precise_timer_done:
+            if (signaled) {
+                current->callback_data.precise.accumulated = 0;
+            } else {
+                current->callback_data.precise.accumulated = (now - start) - delay;
+            }
+            break;
+        }
+        }
```

Comment:
We might need a separate thread for precise timers. Right now, an imprecise timer will block a precise one that's scheduled slightly after it.